BUG: segfault in ndimage.fourier_ellipsoid with length-1 dims

ianthomas23 opened the pull request:
#### Reference issue
Closes gh-12763

#### What does this implement/fix?
Replaces a segmentation fault with a `ValueError`.

The three python functions `ndimage.fourier_ellipsoid`, `fourier_gaussian` and `fourier_uniform` call the C function `NI_FourierFilter` in `ndimage/src/ni_fourier.c`. If the size of any of the dimensions of the supplied input array is less than 2 a segmentation fault occurs in the C code. This PR checks the input array dimension sizes and raises a `ValueError` instead.

The new code is in the C function rather than repeated 3 times in the python functions.

#### Additional information
Test added and docstrings updated.

tylerjereddy reviewed (commented):
Certainly looks promising at first glance with unit testing and doc improvements + CI passing.

peterbell10 commented:
It looks to me like the code is trying to support length-1 dims, it just has a logic bug. Gaussian and Uniform modes always guard their array accesses with `if (params[kk])` but ellipsoid doesn't. 

If you look at the if-else starting on Line 298, it's even trying to handle to the size 1 case. Only it does it by accessing the unallocated array:
```cpp
if (params[hh]) {
   // ..
}
else if (PyArray_DIM(input, hh) > 0) {
    params[hh][0] = 1.0;
}
```

So, I think allocating the array unconditionally when `filter_type == _NI_ELLIPSOID` would be an appropriate fix.

ianthomas23 commented:
@peterbell10 Following your recommendation, I've changed this PR so that `ndimage.fourier_ellipsoid` now supports length-1 dimensions. I've also modified one of each of the `fourier_ellipsoid`, `fourier_gaussian` and `fourier_uniform` tests to show that they all support length-1 dimensions.

Commits squashed to remove changes that I subsequently reverted.

peterbell10 commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -9,7 +9,7 @@
+    @pytest.mark.parametrize('shape', [(32, 16), (31, 15), (1, 10)])
     @pytest.mark.parametrize('dtype, dec',
                              [(numpy.float32, 6), (numpy.float64, 14)])
     def test_fourier_gaussian_real01(self, shape, dtype, dec):
```
Is there a reason you only add this to the real tests?

ianthomas23 replied:
I figured one test for each of the 3 python functions was enough.

peterbell10 commented on scipy/ndimage/src/ni_fourier.c:
```diff
@@ -225,12 +225,10 @@ int NI_FourierFilter(PyArrayObject *input, PyArrayObject* parameter_array,
     }
     for (kk = 0; kk < PyArray_NDIM(input); kk++) {
-        if (PyArray_DIM(input, kk) > 1) {
-            params[kk] = malloc(PyArray_DIM(input, kk) * sizeof(double));
```
I think this might create new issues with length-0 axes. Specifically, `_NI_UNIFORM` on line 270 has:
```c
if (params[hh]) {
	params[hh][0] = 1.0;
    // etc...
}
```
which is invalid if `params[hh]` is a non-null pointer to a 0-sized array.

ianthomas23 replied:
Yes, you're right. I didn't notice as there are no tests for length-0 dimensions, perhaps not surprisingly as there were no tests for length-1 dimensions.

So, should length-0 dimensions be supported? In general this seems like a bad idea but I am new around here.

peterbell10 replied:
I would say it's better to support empty dimensions unless there's a good reason not to. In this case, the result should just be an empty array with the same shape as the input. So long as the parameter initialization doesn't crash it should be fine.

Also note places where the code already checks for `> 0`, so again it's making an effort to support it.

ianthomas23 commented:
Now this PR only changes `fourier_ellipsoid`, the other two functions `fourier_gaussian` and `fourier_uniform` remain as they were.

Added a test for zero-length dimensions. There are quite a few parameter combinations which may be overkill.

peterbell10 commented on scipy/ndimage/tests/test_fourier.py:
```diff
@@ -129,3 +130,16 @@ def test_fourier_ellipsoid_1d_complex(self):
                 assert_array_almost_equal(a, b, decimal=dec)
+
+    @pytest.mark.parametrize('shape', [(0, ), (0, 10), (10, 0)])
+    @pytest.mark.parametrize('dtype, dec',
+                             [(numpy.float32, 5), (numpy.float64, 14),
+                              (numpy.complex64, 5), (numpy.complex128, 14)])
+    @pytest.mark.parametrize('test_func',
+                             [ndimage.fourier_ellipsoid,
+                              ndimage.fourier_gaussian,
+                              ndimage.fourier_uniform])
+    def test_fourier_zero_length_dims(self, shape, dtype, dec, test_func):
```
Nit: The `dec` argument isn't being used.
```suggestion
    @pytest.mark.parametrize('dtype',
                             [numpy.float32, numpy.float64,
                              numpy.complex64, numpy.complex128])
    @pytest.mark.parametrize('test_func',
                             [ndimage.fourier_ellipsoid,
                              ndimage.fourier_gaussian,
                              ndimage.fourier_uniform])
    def test_fourier_zero_length_dims(self, shape, dtype, test_func):
```

ianthomas23 replied:
@peterbell10 `dec` argument removed, thanks for spotting.

peterbell10 reviewed (commented):
Looks good and tests are passing. Just a minor nit-pick.

peterbell10 commented:
Thanks @ianthomas23. Remaining CI failure is just a failed download, so merging.
